compiler-cli/src/dependencies.rs

```diff
@@ -492,38 +516,230 @@ fn resolve_versions<Telem: Telemetry>(
+            }
+        };
+        let _ = version_requirements.insert(name, version);
+    }
+
+    let resolved = dependency::resolve_versions(
         PackageFetcher::boxed(runtime.clone()),
-        mode,
-        config,
-        manifest,
+        provided_packages.clone(),
+        config.name.to_string(),
+        version_requirements.into_iter(),
+        &locked,
     )?;
-    let packages = runtime.block_on(future::try_join_all(
-        resolved
-            .into_iter()
-            .map(|(name, version)| lookup_package(name, version)),
-    ))?;
+
+    let provided_package_requirements = provided_packages
+        .into_iter()
+        .map(|(name, package)| {
+            (
+                name,
+                package.releases[0].requirements.keys().cloned().collect(),
+            )
+        })
+        .collect();
+
+    let packages = runtime.block_on(future::try_join_all(resolved.into_iter().map(
+        |(name, version)| {
+            lookup_package(
+                name,
+                version,
+                &provider_info,
+                &provided_package_requirements,
+            )
+        },
+    )))?;
     let manifest = Manifest {
         packages,
         requirements: config.all_dependencies()?,
     };
     Ok(manifest)
 }
 
-async fn lookup_package(name: String, version: Version) -> Result<ManifestPackage> {
-    let config = hexpm::Config::new();
-    let release = hex::get_package_release(&name, &version, &config, &HttpClient::new()).await?;
-    let manifest = ManifestPackage {
-        name,
-        version,
-        otp_app: Some(release.meta.app),
-        build_tools: release.meta.build_tools,
-        requirements: release.requirements.keys().cloned().collect_vec(),
-        source: ManifestPackageSource::Hex {
-            outer_checksum: Base16Checksum(release.outer_checksum),
-        },
+#[derive(Clone, Eq, PartialEq, Debug)]
+enum ProviderInfo {
+    Git { repo: String, commit: String },
+    Local { path: PathBuf },
+}
+
+fn provide_local_package(
+    package_name: &str,
+    package_path: &Path,
+    info: &mut HashMap<String, ProviderInfo>,
+    provided: &mut HashMap<String, hexpm::Package>,
+) -> Result<hexpm::version::Range> {
+    let canonical_path = package_path
+        .canonicalize()
+        .expect("local package cannonical path");
+    let package_info = ProviderInfo::Local {
+        path: canonical_path.clone(),
     };
-    Ok(manifest)
+
+    // Determine if package has already been walked
+    match info.insert(package_name.to_string(), package_info.clone()) {
+        None => {
+            // No package with this name has been found yet
+            provide_package(package_name, &canonical_path, info, provided)
+        }
+        Some(existing_package_info) => {
+            // A package with this name has already been found
+            // True only if they are both local with the same canonical path, or both git with the same repo and commit
```

Comment:
When can this happen?

Reply:
The most common example would just be cycles, where A has a path dependency to B and B has a path dependency to A. In this case we only want to load each once (to prevent infinite recursion) and let pubgrub explain that theres a cycle with it's fancy errors.

A (much less common) example is where A has path dependencies B and C, and both B and C depend on D, but B and C use different local versions of D. In this case we do need to return an error. This probably will never happen currently but (1) its correct to check for it and (2) possibly more likely with a mixture of git and relative local dependencies.